eth/db/storage.py

```diff
@@ -0,0 +1,477 @@
+        which accounts might have changed storage roots.
+        """
+        self._account_db = account_db
+        self._raw_store_db = db
+        self._account_stores = {}  # type: Dict[Address, AccountStorageDB]
+        self._dirty_accounts = set()  # type: Set[Address]
+
+    def _get_address_store(self, address: Address) -> AccountStorageDB:
+        if address in self._account_stores:
+            store = self._account_stores[address]
+        else:
+            store = AccountStorageDB(self._raw_store_db, self._account_db, address)
+            self._account_stores[address] = store
+        return store
+
+    def _trigger_writeable_storage(self, address: Address) -> None:
+        self._dirty_accounts.add(address)
+
+    def _written_account_stores(self) -> Iterable[Tuple[Address, AccountStorageDB]]:
+        for address in self._dirty_accounts:
+            store = self._account_stores[address]
+            yield address, store
+
+    def get(self, address: Address, slot: int, from_journal: bool=True) -> int:
+        account_store = self._get_address_store(address)
+        return account_store.get(slot, from_journal)
+
+    def set(self, address: Address, slot: int, value: int) -> None:
+        account_store = self._get_address_store(address)
+        if address not in self._dirty_accounts:
+            self._trigger_writeable_storage(address)
+        account_store.set(slot, value)
+
+    def delete(self, address: Address) -> None:
+        account_store = self._get_address_store(address)
+        if address not in self._dirty_accounts:
+            self._trigger_writeable_storage(address)
+        account_store.delete()
+
+    def record(self) -> UUID:
+        changeset_id = uuid4()
+        for _, store in self._written_account_stores():
+            store.record(changeset_id)
+        return changeset_id
+
+    def discard(self, changeset: UUID) -> None:
+        for _, store in self._written_account_stores():
+            store.discard(changeset)
+
+    def commit(self, changeset: UUID) -> None:
+        for _, store in self._written_account_stores():
+            store.commit(changeset)
+
+    def make_storage_roots(self) -> Tuple[Tuple[Address, Hash32], ...]:
+        self.logger.debug2('make storage roots for all accounts')
+        for _, store in self._written_account_stores():
+            store.make_storage_root()
+        return self._get_changed_roots()
+
+    @to_tuple
+    def _get_changed_roots(self) -> Iterable[Tuple[Address, Hash32]]:
+        # list all the accounts that were changed, and their new storage roots
+        for address, store in self._written_account_stores():
+            if store.has_changed_root:
+                yield address, store.get_changed_root()
+
+    def persist(self) -> Tuple[Tuple[Address, Hash32], ...]:
+        self.logger.debug2('persisting storage roots')
+
+        # must generate changed_roots before persisting
+        changed_roots = self._get_changed_roots()
```

Comment:
Looking at this some more, I was surprised that this didn't work!

```python
    storage_db.set(ADDRESS, 1, 1)
    storage_db.make_storage_roots()
    storage_db.persist()  # throws `ValidationError`
```

In retrospect, I now see that the error message was telling me to do this:

```python
    storage_db.set(ADDRESS, 1, 1)

    changed_roots = storage_db.make_storage_roots()
    for addr, storage_root in changed_roots:
        account_db.set_storage_root(addr, storage_root)
    storage_db.persist()
```

But reading the code wasn't enough for me to figure this out. Eventually I had to look at the tests. I think the reason `make_storage_roots()` can't be merged into `persist()` is that in frontier we want to make a state root for each receipt but don't necessarily want to persist each time. However, given that `StorageDB` is already aware of `AccountDB`, what do you think of having `StorageDB.make_storage_roots()` do the `set_storage_root()` part itself? I think that would lead to more straightforward code, and save the programmer from having to remember something `StorageDB` already knows.

Reply:
Oh yeah, somehow a change to that comment didn't make it in here... :eyes:  yup, it was in the next patch-list.

`AccountStorageDB` actually enforces that there is no changed data since the last root was calculated, on each `AccountStorageDB.persist()` call, so I'll just remove the comment. I think the docstring on `persist()` covers it.

> Is there any harm in turning this into a call to `make_storage_roots`? Or maybe it could validate that the state roots have been generated, or maybe `make_storage_roots()` and `persist()` could be combined. Currently there's only one caller so the knowledge that one method must be called before the other isn't exactly distributed but it still feels like it belongs in `StorageDB`.

`VMState` always needs to know the output of `make_storage_roots()` so it can update the `AccountDB`. We could have persist return the updated storage roots, but it feels easy to miss the necessary step of capturing the storage roots coming out of `persist()`. Instead requiring that the caller explicitly handle the new storage roots before persisting seemed the cleanest. Especially since the storage roots are sometimes generated without persisting.

Reply:
> VMState always needs to know the output of make_storage_roots() so it can update the AccountDB.

Do you mean in `BaseState.make_state_root()`? It looks like there it takes the state roots and immediately gives them to `_account_db` and then they go out of scope. Since `StorageDB` already has a reference to `AccountDB` I think it could become easier to work with by calling `set_storage_root()` itself in `make_storage_roots`.